jumanji/environments/combinatorial/routing/generation_methods_v1_uo.py

```diff
@@ -0,0 +1,572 @@
+            self.filled = True
+            self.populate_grid()
+            self.solved_board = self.grid.layout.copy()
+            self.remove_routes()
+            return self.solved_board, self.empty_board, self.wires
+
+    def fill_board_with_clipping(self, num_clips: int = 2, method: str = 'bends', verbose: Optional[bool] = False) -> \
+            Tuple[np.ndarray, np.ndarray, int]:
+        """Fills the board, clips the wires and then tries to fill again.
+        Args:
+            num_clips (int): number of wires to remove
+            method (str): method to remove the wires
+            verbose (bool): if True, prints the progress
+        Returns:
+            tuple[np.array, np.array, int]: the board, the solved board and the number of filled wires
+        """
+        self.place_wires(verbose=verbose)
+        self.clip(method, num_clips)
+
+        # Try to fill again
+        self.place_wires(verbose=verbose)
+
+        return self.return_boards(verbose=verbose)
+
+    def return_empty_board(self) -> np.ndarray:
+        """Returns the empty board with heads and targets but not wires.
+        Returns:
+            np.array: the empty board
+        """
+        if self.filled:
+            return self.empty_board
+        else:
+            return self.unsolved_board
+
+    def return_solved_board(self) -> np.ndarray:
+        """Returns the solved board with heads, targets and wires.
+        Returns:
+            np.array: the solved board
+        """
+        if self.filled:
+            return self.solved_board
+        else:
+            return self.partial_board
+
+    def fill_clip_fill(self, num_clips: Union[int, List[int]], methods: Union[str, List[str]],
+                       num_loops: Optional[int] = 1, verbose: Optional[bool] = False) -> \
+            Tuple[np.ndarray, np.ndarray, int]:
+        """ Performs a number of fill, clip, fill loops.
+        Args:
+            num_clips (int): number of wires to remove
+            methods (str): method to remove the wires
+            num_loops (int): number of loops to perform
+            verbose (bool): if True, prints the progress
+        Returns:
+            tuple[np.array, np.array, int]: the board, the solved board and the number of filled wires
+        """
+        if isinstance(num_clips, int):
+            num_clips = [num_clips] * num_loops
+        if isinstance(methods, str):
+            methods = [methods] * num_loops
+        assert len(num_clips) == len(methods), "Number of clips and methods must be the same."
+        if verbose:
+            print(f"Performing {num_loops} loops of fill-clip-fill.")
+        for i in range(num_loops):
+            self.place_wires(verbose=verbose)
+            self.clip(methods[i], num_clips[i])
+
+        # Perform the final fill
+        self.place_wires(verbose=verbose)
+
+        # Return the boards
+        return self.return_boards(verbose=verbose)
+
+
+if __name__ == '__main__':
+    # Example usage
```

Comment:
Great example usage, I think I would be able to pick everything I needed from here. Let's talk about how / if you want to benchmark these against each other for Thursday?

Reply:
Idea so I dont forget: maybe we can make classes that inherit from this one that return the board with the appropriate methods if we dont want to split out into separate classes